source/extensions/common/wasm/context.cc

```diff
@@ -854,20 +854,7 @@ BufferInterface* Context::getBuffer(WasmBufferType type) {
 
 void Context::onDownstreamConnectionClose(CloseType close_type) {
   ContextBase::onDownstreamConnectionClose(close_type);
-  downstream_closed_ = true;
-  // Call close on TCP connection, if upstream connection closed or there was a failure seen in
-  // this connection.
-  if (upstream_closed_ || getRequestStreamInfo()->hasAnyResponseFlag()) {
-    onCloseTCP();
-  }
-}
-
-void Context::onUpstreamConnectionClose(CloseType close_type) {
-  ContextBase::onUpstreamConnectionClose(close_type);
-  upstream_closed_ = true;
-  if (downstream_closed_) {
-    onCloseTCP();
-  }
+  onCloseTCP();
```

Comment:
This whole change could be basically minimized to this (i.e. on downstream close, close TCP context).

I don't think we can remove upstream close event, since it's a breaking change, or that we even need to remove it in order to fix the issue at hand.

Reply:
1. Who exactly are we breaking? I don't think anyone relies on this callback or they would definitely complain about it before me.

2. What's the difference between upstream close downstream local close? What does upstream close mean in case of direct response?

Reply:
1. I don't know, we don't have an exhaustive list of everybody using Proxy-Wasm extensions. Notably, this seems to be only broken when using clusters with TLS transport socket (see: #13856).

2. I think you're confusing downstream/upstream with remote/local close events:
    - downstream local close means that the downstream connection was closed by Envoy,
    - downstream remote close means that the downstream connection was closed by the client,
    - upstream local close means that the upstream connection was closed by Envoy,
    - upstream remote close means that the upstream connection was closed by the backend,

In any case, this bug is not unique to Wasm, and I think the proper fix should emit upstream connection close event (again, see: #13856).

Reply:
OK, #13856 is not the root cause for #13806, but it results in the same leak.

For #13806 the issue is that the upstream close event won't be triggered if the connection to upstream was never established (e.g. connection timeout, connection refused), but once #13856 is fixed, then this event is always triggered if the connection to upstream was established, so I don't see the point in removing it.

Like I said originally, this PR should be reduced to "on downstream close event, destroy context" (ideally, we should keep waiting for upstream close if we ever transmitted data in either direction), but without removing the upstream close event.

Reply:
I can reproduce this without TLS.

Reply:
Right, as mentioned in the previous comment, there are 2 sources of the memory leak: one is TLS clusters, the other one is connection failures (connection timeout, connection refused, etc).

Reply:
What about direct response or black hole traffic?

Reply:
I imagine that direct response works fine, since it probably sets `end_stream`, and the black hole falls under connection failures... but it would be good to verify both.

Reply:
I think it would be confusing to have an upstream connection event where there is no upstream connection (as in direct response).

Reply:
This is debatable. After all, the plugin is receiving "upstream data" from the direct response as well. The way I look at it, from the point of the plugin, it doesn't matter whether operator configured Envoy with `tcp_proxy` or `direct_response` as the next filter, since both are "upstream" of the plugin itself.

Again, I agree that we should destroy context on the downstream connection close event alone, if there is no upstream connection, but I disagree with completely removing the upstream connection close event, especially since removing it isn't necessary in order to fix the memory leak.

Also, all the connection failures should be covered by the `hasAnyResponseFlag()` code path here:
https://github.com/envoyproxy/envoy/blob/d5e8634edf26dab65218e42f2acfe7c3dc95b72d/source/extensions/common/wasm/context.cc#L860

Under what conditions is the downstream connection closed without any response flag being set? Now that the TLS connection close detection is fixed (#13858), I cannot find any case without response flags.

Reply:
I don't have an easy way to detect the conditions since I reproduce with a stress test.

Reply:
Could you print response flags or error if missing? Alternatively, do you have repro steps for the stress that I could play with locally?

Reply:
I'm not sure how to easily debug this. The issue is that either callback is not called or response flags are empty. It's hard to detect absence.

BTW I'm still not convinced about "upstream data". What does that even mean in the upstream network filter? We're using network filters on clusters in istio.

Reply:
> I'm not sure how to easily debug this. The issue is that either callback is not called or response flags are empty. It's hard to detect absence.

Could you share the steps to reproduce so that I can try to debug it?

> BTW I'm still not convinced about "upstream data". What does that even mean in the upstream network filter? We're using network filters on clusters in istio.

It might work by accident, but Proxy-Wasm plugins don't officially support "upstream network filters", and I suspect that some things will be broken because of flipped direction and related checks. I'm also pretty sure there are no tests for that use case. Please open an issue if you want to use them as "upstream network filters".

Reply:
1. Add this diff:
```c++
  ::Wasm::Common::RequestInfo* request_info);
   void deleteFromRequestQueue(uint32_t context_id);

+  int64_t log_count = 0;
+  int64_t connection_count = 0;
+
  protected:
   const std::vector<MetricTag>& defaultTags();
   const std::vector<MetricFactory>& defaultMetrics();
@@ -333,6 +336,8 @@ class PluginContext : public Context {
     if (request_info_.request_protocol == ::Wasm::Common::Protocol::TCP) {
       request_info_.tcp_connections_closed++;
     }
+
+    rootContext()->log_count++;
     rootContext()->report(request_info_, true);
   };

@@ -361,6 +366,7 @@ class PluginContext : public Context {
     request_info_.request_protocol = ::Wasm::Common::Protocol::TCP;
     request_info_.tcp_connections_opened++;
     rootContext()->addToRequestQueue(id(), &request_info_);
+    rootContext()->connection_count++;
     return FilterStatus::Continue;
   }
```

2. Start server: `fortio server`.
3. Start envoy with above config, concurrency 1.
4. Generate load: `fortio load -qps 10000 -c 100 -t 10s localhost:8001`
5. Terminate ^C server, restart it, and keep doing it until load finishes.
6. Observe mismatched counts.

Reply:
Filed #13929 .

Reply:
Thanks for the instructions, I was able to reproduce the issue. There are mid-stream I/O errors that are silently dropped in transport sockets, see: #13939. Once the core issue is fixed (my attempt in #13941), the counters always match.

Reply:
I don't have high confidence that there's some other issue lurking about error propagation. If this happened under trivial config, it's very likely that more complex config can cause the same leak. We should not expect that upstream connection events are correctly emitted.